README.md
# Rendology
[![Docs Status](https://docs.rs/rendology/badge.svg)](https://docs.rs/rendology)
[![license](http://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/leod/rendology/blob/master/LICENSE)
[![Crates.io](https://img.shields.io/crates/v/rendology.svg)](https://crates.io/crates/rendology)

![Ultimate Scale screenshot](https://leod.github.io/assets/rendology_hdr_0_4.png)

Rendology is a [Glium](https://github.com/glium/glium)-based 3D rendering library. It is not, however, a game engine.

:warning: Rendology is in a very early stage; it is undocumented and subject to frequent changes. Using this crate is not (yet) recommended! :warning:

There's a [blog post](https://leod.github.io/rust/gamedev/rendology/2019/12/13/introduction-to-rendology.html) on Rendology's architecture.

## Features
Rendology features amateur-grade implementations of:

- Shadow mapping
- Deferred shading
- A glow effect
- FXAA
- Instanced rendering

Each of the rendering effects can be turned on and off separately.

Rendology makes it easy to integrate custom scene shaders into the rendering pipeline.


CHANGELOG.md
# Changelog
## Version 0.5.0
- Remove dependency on `genmesh` and `obj` crates ([#14](https://github.com/leod/rendology/pull/14))
- Base shadow map core on given fragment core ([#15](https://github.com/leod/rendology/pull/15))
- Move associated types of `SceneCore` into new trait `CoreInput` ([#16](https://github.com/leod/rendology/pull/16))
- Initial implementation of particles ([#17](https://github.com/leod/rendology/pull/17), [#19](https://github.com/leod/rendology/pull/19))
- Add a shader transform for transparency with dithering ([#20](https://github.com/leod/rendology/pull/20))
- Change shader compilation so that output assignments can be evaluated before bodies ([#20](https://github.com/leod/rendology/pull/20))
- Performance tuning ([#21](https://github.com/leod/rendology/pull/21))
- Increase brightness of glowing objects even if blur is disabled ([#22](https://github.com/leod/rendology/pull/22))

## Version 0.4.1 (2019-12-17)
- Fix bug in shadow mapping on Intel GPUs ([#12](https://github.com/leod/rendology/pull/12))

## Version 0.4.0 (2019-12-12)
- `Pipeline::start_frame` now takes the clear color ([#6](https://github.com/leod/rendology/pull/6))
- Implement basic line rendering ([#7](https://github.com/leod/rendology/pull/7))
- Allow plain scene pass after postprocessing ([#8](https://github.com/leod/rendology/pull/8))
- Implement simple shadow smoothing ([#9](https://github.com/leod/rendology/pull/9))
- Rename `V_POSITION` to `V_POS` ([#10](https://github.com/leod/rendology/pull/10))
- Revamp shader input/output definitions, breaking API ([#11](https://github.com/leod/rendology/pull/11))

## Version 0.3.0 (2019-12-10)
- Allow passing draw parameters to shadow pass ([#3](https://github.com/leod/rendology/pull/3))
- Overhaul `ToUniforms` to allow lifetimes ([#3](https://github.com/leod/rendology/pull/3))
- Add textured cube example ([#3](https://github.com/leod/rendology/pull/3))
- Fix bug where glow output texture was not set ([#3](https://github.com/leod/rendology/pull/3))
- `impl_instance_input` and `impl_uniform_input` now take the actual rust types ([#3](https://github.com/leod/rendology/pull/3))
- Reworked macro syntax ([#4](https://github.com/leod/rendology/pull/4))
- More explicit names for built-in uniforms ([#4](https://github.com/leod/rendology/pull/4))
- Make more shading parameters configurable ([#5](https://github.com/leod/rendology/pull/5))

## Version 0.2.0 (2019-12-06)
- Replace `ToVertex` trait by `InstanceInput`
- Allow specifying the `InstancingMode` for scene passes
- Implement uniform-based drawing of `RenderList`

## Version 0.1.0 (2019-12-06)
- Initial version


src/object.rs
use std::path::Path;

use glium::{self, implement_vertex};
use log::info;
use num_derive::{FromPrimitive, ToPrimitive};

pub use crate::CreationError;

#[derive(Copy, Clone, PartialEq, Eq, Debug, FromPrimitive, ToPrimitive)]
pub enum Object {
    Triangle,
    Quad,
    Cube,
    Sphere,

    LineX,
    LineY,
    LineZ,

    TessellatedCube,
    TessellatedCylinder,

    /// Counter of the number of objects
    NumTypes,
}

#[derive(Copy, Clone, Debug)]
pub struct Vertex {
    pub position: [f32; 3],
    pub normal: [f32; 3],
}

implement_vertex!(Vertex, position, normal);

pub enum IndexBuffer {
    IndexBuffer(glium::index::IndexBuffer<u32>),
    NoIndices(glium::index::NoIndices),
}

impl IndexBuffer {
    pub fn draw<'a, V, U, S>(
        &self,
        vertex_buffer: V,
        program: &glium::Program,
        uniforms: &U,
        draw_parameters: &glium::DrawParameters,
        target: &mut S,
    ) -> Result<(), glium::DrawError>
    where
        V: glium::vertex::MultiVerticesSource<'a>,
        U: glium::uniforms::Uniforms,
        S: glium::Surface,
    {
        match &self {
            IndexBuffer::IndexBuffer(buffer) => {
                target.draw(vertex_buffer, buffer, program, uniforms, draw_parameters)
            }
            IndexBuffer::NoIndices(buffer) => {
                target.draw(vertex_buffer, buffer, program, uniforms, draw_parameters)
            }
        }
    }
}

pub struct ObjectBuffers<V: Copy> {
    pub vertex_buffer: glium::VertexBuffer<V>,
    pub index_buffer: IndexBuffer,
}

impl ObjectBuffers<Vertex> {
    pub fn from_slices<F: glium::backend::Facade>(
        facade: &F,
        primitive_type: glium::index::PrimitiveType,
        positions: &[[f32; 3]],
        normals: &[[f32; 3]],
        indices: &[u32],
    ) -> Result<ObjectBuffers<Vertex>, CreationError> {
        let vertices = positions
            .iter()
            .zip(normals.iter())
            .map(|(&p, &n)| Vertex {
                position: p,
                normal: n,
            })
            .collect::<Vec<_>>();

        Ok(ObjectBuffers {
            vertex_buffer: glium::VertexBuffer::new(facade, &vertices)?,
            index_buffer: IndexBuffer::IndexBuffer(glium::IndexBuffer::new(
                facade,
                primitive_type,
                indices,
            )?),
        })
    }

    pub fn load_wavefront<F: glium::backend::Facade>(
        facade: &F,
        path: &Path,
    ) -> Result<ObjectBuffers<Vertex>, CreationError> {
        info!("Loading Wavefront .OBJ file: `{}'", path.display());

        // As in:
        // https://github.com/glium/glium/blob/master/examples/support/mod.rs

        let data = obj::Obj::load(path).unwrap();

        let mut vertices = Vec::new();

        for object in data.objects.iter() {
            for polygon in object.groups.iter().flat_map(|g| g.polys.iter()) {
                match polygon {
                    genmesh::Polygon::PolyTri(genmesh::Triangle {
                        x: v1,
                        y: v2,
                        z: v3,
                    }) => {
                        for v in [v1, v2, v3].iter() {
                            let position = data.position[v.0];
                            let normal = v.2.map(|index| data.normal[index]);

                            let normal = normal.unwrap_or([0.0, 0.0, 0.0]);

                            vertices.push(Vertex { position, normal })
                        }
                    }
                    _ => unimplemented!(),
                }
            }
        }

        let vertex_buffer = glium::VertexBuffer::new(facade, &vertices).unwrap();
        let primitive_type = glium::index::PrimitiveType::TrianglesList;
        let index_buffer = IndexBuffer::NoIndices(glium::index::NoIndices(primitive_type));

        Ok(ObjectBuffers {
            vertex_buffer,
            index_buffer,
        })
    }
}

impl Object {
    #[rustfmt::skip]
    pub fn create_buffers<F: glium::backend::Facade>(
        self, facade: &F
    ) -> Result<ObjectBuffers<Vertex>, CreationError> {
        match self {
            Object::Triangle => {
                let positions = vec![
                    [0.0, -0.5, 0.0],
                    [0.0,  0.5, 0.0],
                    [1.0,  0.0, 0.0],
                ];

                let normals = vec![
                    [0.0, 0.0, -1.0],
                    [0.0, 0.0, -1.0],
                    [0.0, 0.0, -1.0],
                ];

                let indices = vec![0, 1, 2];

                ObjectBuffers::from_slices(
                    facade,
                    glium::index::PrimitiveType::TrianglesList,
                    &positions,
                    &normals,
                    &indices,
                )
            }
            Object::Quad => {
                let positions = vec![
                    [0.0, 0.0, 0.0],
                    [1.0, 0.0, 0.0],
                    [1.0, 1.0, 0.0],
                    [0.0, 1.0, 0.0],
                ];

                let normals = vec![
                    [0.0, 0.0, 1.0],
                    [0.0, 0.0, 1.0],
                    [0.0, 0.0, 1.0],
                    [0.0, 0.0, 1.0],
                ];

                let indices = vec![
                    0, 1, 2,
                    2, 3, 0,
                ];

                ObjectBuffers::from_slices(
                    facade,
                    glium::index::PrimitiveType::TrianglesList,
                    &positions,
                    &normals,
                    &indices,
                )
            }
            Object::Cube => {
                ObjectBuffers::from_slices(
                    facade,
                    glium::index::PrimitiveType::TrianglesList,
                    CUBE_POSITIONS,
                    CUBE_NORMALS,
                    CUBE_INDICES,
                )
            }
            Object::Sphere => {
                // For reference: http://www.songho.ca/opengl/gl_sphere.html

                let mut positions = Vec::new();
                let mut normals = Vec::new();
                let mut indices = Vec::new();

                let radius = 0.5;
                let num_stacks = 10;
                let num_sectors = 10;

                let sector_step = 2.0 * std::f32::consts::PI / num_sectors as f32;
                let stack_step = std::f32::consts::PI / num_stacks as f32;

                for i in 0..=num_stacks {
                    // Phi goes from pi/2 (top) to -pi/2 (bottom)
                    let phi = std::f32::consts::PI / 2.0 - i as f32 * stack_step;

                    for j in 0..=num_sectors {
                        // Theta goes from 0 to 2*pi (around the sphere)
                        let theta = j as f32 * sector_step;

                        let x = phi.cos() * theta.cos();
                        let y = phi.cos() * theta.sin();
                        let z = phi.sin();

                        positions.push([x * radius, y * radius, z * radius]);
                        normals.push([x, y, z]);
                    }
                }

                for i in 0..num_stacks {
                    // Beginning of current stack
                    let k_1 = i * (num_sectors + 1);

                    // Beginning of next stack
                    let k_2 = k_1 + (num_sectors + 1);

                    for j in 0..num_sectors {
                        if i != 0 {
                            indices.push(k_1 + j);
                            indices.push(k_2 + j);
                            indices.push(k_1 + j + 1);
                        }

                        if i + 1 != num_stacks {
                            indices.push(k_1 + j + 1);
                            indices.push(k_2 + j);
                            indices.push(k_2 + j + 1);
                        }
                    }
                }

                ObjectBuffers::from_slices(
                    facade,
                    glium::index::PrimitiveType::TrianglesList,
                    &positions,
                    &normals,
                    &indices,
                )
            }
            Object::LineX => {
                ObjectBuffers::from_slices(
                    facade,
                    glium::index::PrimitiveType::LinesList,
                    &[[0.0, 0.0, 0.0], [1.0, 0.0, 0.0]],
                    &[[0.0, 0.0, 1.0], [0.0, 0.0, 1.0]],
                    &[0, 1],
                )
            }
            Object::LineY => {
                ObjectBuffers::from_slices(
                    facade,
                    glium::index::PrimitiveType::LinesList,
                    &[[0.0, 0.0, 0.0], [0.0, 1.0, 0.0]],
                    &[[0.0, 0.0, 1.0], [0.0, 0.0, 1.0]],
                    &[0, 1],
                )
            }
            Object::LineZ => {
                ObjectBuffers::from_slices(
                    facade,
                    glium::index::PrimitiveType::LinesList,
                    &[[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]],
                    &[[1.0, 0.0, 0.0], [1.0, 0.0, 0.0]],
                    &[0, 1],
                )
            }
            Object::TessellatedCube => {
                let mut positions = Vec::new();
                let mut normals = Vec::new();
                let mut indices = Vec::new();

                // Number of subdivisions along the x axis
                let n = 64;

                for i in 0..n {
                    // Add one x-slice of the cube
                    let x_offset = i as f32 / n as f32 - 0.5;

                    for (&position, &normal) in CUBE_POSITIONS.iter().zip(CUBE_NORMALS.iter()) {
                        positions.push([x_offset + position[0] / n as f32, position[1], position[2]]);
                        normals.push(normal);
                    }

                    for &index in CUBE_INDICES {
                        indices.push((CUBE_POSITIONS.len() * i) as u32 + index);
                    }
                }

                ObjectBuffers::from_slices(
                    facade,
                    glium::index::PrimitiveType::TrianglesList,
                    &positions,
                    &normals,
                    &indices,
                )
            }
            Object::TessellatedCylinder => {
                let mut positions = Vec::new();
                let mut normals = Vec::new();
                let mut indices = Vec::new();

                // Number of subdivisions along the x axis
                let n = 32;

                // Number of subdivisions along the angle
                let m = 8;

                // Add positions and normals
                for i in 0..=n {
                    let x = i as f32 / n as f32 - 0.5;

                    // Add one x-slice of the cylinder
                    for j in 0..m {
                        // Add one stripe of the cylinder
                        let theta = j as f32 / m as f32 * 2.0 * std::f32::consts::PI;

                        let y = theta.sin();
                        let z = theta.cos();
                        positions.push([x, y, z]);
                        normals.push([0.0, y, z]);
                    }
                }

                // Add triangles
                for i in 0..n {
                    for j in 0..m {
                        indices.push((i + 0) * m + ((j + 0) % m));
                        indices.push((i + 1) * m + ((j + 0) % m));
                        indices.push((i + 0) * m + ((j + 1) % m));

                        indices.push((i + 0) * m + ((j + 1) % m));
                        indices.push((i + 1) * m + ((j + 0) % m));
                        indices.push((i + 1) * m + ((j + 1) % m));
                    }
                }

                ObjectBuffers::from_slices(
                    facade,
                    glium::index::PrimitiveType::TrianglesList,
                    &positions,
                    &normals,
                    &indices,
                )
            }
            Object::NumTypes => panic!("Object::NumTypes cannot be instantiated!"),
        }
    }
}

#[rustfmt::skip]
pub const CUBE_POSITIONS: &[[f32; 3]] = &[
    // Front
    [-0.5, -0.5,  0.5],
    [ 0.5, -0.5,  0.5],
    [ 0.5,  0.5,  0.5],
    [-0.5,  0.5,  0.5],

    // Right
    [ 0.5,  0.5, -0.5],
    [ 0.5,  0.5,  0.5],
    [ 0.5, -0.5,  0.5],
    [ 0.5, -0.5, -0.5],

    // Back
    [ 0.5, -0.5, -0.5],
    [-0.5, -0.5, -0.5],
    [-0.5,  0.5, -0.5],
    [ 0.5,  0.5, -0.5],

    // Left
    [-0.5, -0.5, -0.5],
    [-0.5, -0.5,  0.5],
    [-0.5,  0.5,  0.5],
    [-0.5,  0.5, -0.5],

    // Top
    [-0.5,  0.5,  0.5],
    [ 0.5,  0.5,  0.5],
    [ 0.5,  0.5, -0.5],
    [-0.5,  0.5, -0.5],

    // Bottom
    [-0.5, -0.5, -0.5],
    [ 0.5, -0.5, -0.5],
    [ 0.5, -0.5,  0.5],
    [-0.5, -0.5,  0.5],
];

#[rustfmt::skip]
pub const CUBE_NORMALS: &[[f32; 3]] = &[
    // Front
    [ 0.0,  0.0,  1.0],
    [ 0.0,  0.0,  1.0],
    [ 0.0,  0.0,  1.0],
    [ 0.0,  0.0,  1.0],

    // Right
    [ 1.0,  0.0,  0.0],
    [ 1.0,  0.0,  0.0],
    [ 1.0,  0.0,  0.0],
    [ 1.0,  0.0,  0.0],

    // Back
    [ 0.0,  0.0, -1.0],
    [ 0.0,  0.0, -1.0],
    [ 0.0,  0.0, -1.0],
    [ 0.0,  0.0, -1.0],

    // Left
    [-1.0,  0.0,  0.0],
    [-1.0,  0.0,  0.0],
    [-1.0,  0.0,  0.0],
    [-1.0,  0.0,  0.0],

    // Top
    [ 0.0,  1.0,  0.0],
    [ 0.0,  1.0,  0.0],
    [ 0.0,  1.0,  0.0],
    [ 0.0,  1.0,  0.0],

    // Bottom
    [ 0.0, -1.0,  0.0],
    [ 0.0, -1.0,  0.0],
    [ 0.0, -1.0,  0.0],
    [ 0.0, -1.0,  0.0],
];

#[rustfmt::skip]
pub const CUBE_INDICES: &[u32] = &[
    // Front
    0, 1, 2, 0, 2, 3,

    // Right
    4, 5, 6, 4, 6, 7,

    // Back
    8, 9, 10, 8, 10, 11,

    // Left
    12, 13, 14, 12, 14, 15,

    // Top
    16, 17, 18, 16, 18, 19,

    // Bottom
    20, 21, 22, 20, 22, 23,
];
